substrate/network-libp2p/src/network_state.rs

```diff
@@ -548,7 +551,8 @@ impl NetworkState {
 		let node_is_reserved = self.reserved_peers.read().contains(&infos.id);
 		if !node_is_reserved {
 			if self.reserved_only.load(atomic::Ordering::Relaxed) ||
-				num_open_connections >= self.max_peers
+				num_open_connections.total >= self.max_peers ||
+				num_open_connections.incoming >= self.max_incoming_peers
```

Comment:
This code is also reached when we dial. You should use `endpoint` to determine whether we are dialing or listening.